### PR TITLE
split `ParserGridDistribution` into `hpp/cpp` file

### LINKED ISSUE #2898: Boost Error in ResourceLog
In combination with:
* GCC 7.3.1
* CUDA 9.2.88
* Boost 1.68.0 (C++11 build) OR Boost 1.65.1
* PIConGPU 0.4.0
* centos 7

I have checked the issue here:
https://github.com/ComputationalRadiationPhysics/picongpu/issues/2714#issue-367113446
But, there  is still such err in ResourceLog.hpp:
```
/trinity/home/b.lei/pic/src/blei/picongpu/include/pmacc/../picongpu/plugins/ResourceLog.hpp(115): here
/trinity/shared/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/stl_pair.h(331): error: identifier "std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_S_empty_rep_storage" is undefined in device code

/trinity/shared/opt/rh/devtoolset-7/root/usr/include/c++/7/bits/stl_pair.h(331): error: identifier "std::basic_string<char, std::char_traits<char>, std::allocator<char> >::_Rep::_S_empty_rep_storage" is undefined in device code

2 errors detected in the compilation of "/tmp/tmpxft_0000926d_00000000-6_main.cpp1.ii".
CMake Error at picongpu_generated_main.cpp.o.Release.cmake:279 (message):
  Error generating file
  /trinity/home/b.lei/pic/build/CMakeFiles/picongpu.dir//./picongpu_generated_main.cpp.o
```

### PR DESCRIPTION
**Pull request by psychocoderHPC**
hopefully, fix #2898

- Split the grid parser into an interface file `hpp` and the implementation `cpp` to avoid that boost regex is parsed by nvcc. Boost regex is very often the root of compile issues because nvcc is not able to split the host and device part correctly.
- Define `abortWithError` static to allow the usage in multiple object files without creating the linker error ` multiple definitions of ...
- remove `boost::regex` by `std::regex`
- remove `boost::lexical_cast`

### CODE REVIEW

**Comment by psychocoderHPC** [comment]
@ax3l The suggested refactoring to c++ regex you mentioned https://github.com/ComputationalRadiationPhysics/picongpu/issues/2898#issuecomment-465128305 can be added in a follow-up PR.

**Comment by ax3l** [comment]
@psychocoderHPC thx! If you ask me, add the C++11 regex in the same PR. We have to backport both anyway.

**Comment by psychocoderHPC** [comment]
I do not think that this PR fill fix #2899 fully. I will also have a look to the resource manager.

**Comment by sbastrakov** [comment]
@psychocoderHPC is it not possible to make `abortWithError` inline and non-static?

**Comment by ax3l** [comment]
`inline` or `static` for global functions only instruct the linker to either always take the local symbol or assume they are all the same. either way, it will work here:

https://stackoverflow.com/questions/185624/static-variables-in-an-inlined-function

**Comment by ax3l** [comment]
> I do not think that this PR fill fix #2899 fully. I will also have a look to the resource manager.

Yep, I just would like to have the `gridDist` parser changes in one place :)

**Comment by sbastrakov** [comment]
@ax3l sure, for our case both `static` and `inline` would do the job. Actually, from a purist point of view it is even better to put a function in an anonymous namespace (yes, even in a header file). But practically (even for me) seems like too much typing for a single function.

**Comment by ax3l** [comment]
Agreed, an anonymous namespace is actually the best option.

**Comment by psychocoderHPC** [comment]
I updated the code and removed the boost dependencies for the parser.
Please be not confused by the strange escaping e.g. ` R"(\{.*})",`. 
Only `{` must be escaped (this behavior cost me over 1h of my lifetime)

**Comment by psychocoderHPC** [comment]
This PR is still WIP because I am checking if we can remove the dependency to `boost::regex` now.

**Comment by ax3l** [comment]
It seems that `boost::iostreams` and `boost::log` depend on `boost::regex`. `boost::program_options` does surprisingly not.

https://gitlab.kitware.com/cmake/cmake/blob/v3.13.4/Modules/FindBoost.cmake#L845-878

File: include/picongpu/initialization/ParserGridDistribution.cpp [hunk 1]
```diff
@@ -25,8 +25,7 @@
 #include <string>   // std::string
 #include <iterator> // std::distance
 
-#include <boost/regex.hpp>
-#include <boost/lexical_cast.hpp>
+#include <regex>
```
Comment (ax3l) [inline include/picongpu/initialization/ParserGridDistribution.cpp]: bye, byte GCC 4.8 :stars: 
(`INSTALL.md` already lists GCC 4.9+, so that's fine.)

**Comment by psychocoderHPC** [comment]
> It seems that `boost::iostreams` and `boost::log` depend on `boost::regex`. `boost::program_options` does surprisingly not, not even transitively.
> 
> https://gitlab.kitware.com/cmake/cmake/blob/v3.13.4/Modules/FindBoost.cmake#L845-878
> 
> https://github.com/ComputationalRadiationPhysics/picongpu/blob/0.4.3/INSTALL.rst#boost
> https://github.com/ComputationalRadiationPhysics/picongpu/blob/0.4.3/include/picongpu/CMakeLists.txt#L168-L169
> https://github.com/ComputationalRadiationPhysics/picongpu/blob/0.4.3/include/pmacc/PMaccConfig.cmake#L308

@ax3l We do not use boost::log or boost::iostream, I tested now also if we can remove the dependency and it looks good.

**Comment by psychocoderHPC** [comment]
@ax3l I removed in the last commit the requirement to have boost::regex. But since splash2txt still need it I only marked boost::regex in the documentation as optional.

I will put on my todo list to remove boost::regex from splash2txt too.

**Comment by ax3l** [comment]
looks good to me, runtime tests were successful?

**Comment by psychocoderHPC** [comment]
Yes I tested different distributions.

**Pull request merged**